File: src/main/kotlin/io/github/hxxniverse/hobeaktown/feature/wasteland/WasteLandFeature.java
package io.github.hxxniverse.hobeaktown.feature.wasteland;

import io.github.hxxniverse.hobeaktown.util.base.BaseFeature;
import org.bukkit.Location;
import org.bukkit.Material;
import org.bukkit.block.Block;
import org.bukkit.inventory.Inventory;
import org.bukkit.inventory.ItemStack;
import org.bukkit.plugin.java.JavaPlugin;
import org.jetbrains.annotations.NotNull;

import java.util.*;

public class WasteLandFeature implements BaseFeature {
    private static WasteLandFeature instance;
    private final Map<Integer, ItemStack> guiItemMap = new HashMap<>();
    private final Map<Location, Material> brushedBlocks = new HashMap<>();

    private WasteLandGui gui;

    @Override
    public void onEnable(@NotNull JavaPlugin plugin) {
        instance = this;
        gui = new WasteLandGui();
        loadData();
    }

    @Override
    public void onDisable(@NotNull JavaPlugin plugin) {
        saveData();

        // 블럭이 바뀌는 3분 동안 서버(플러그인)가 종료되는 경우 즉시 변경 실행
        Block block;
        for(Map.Entry entry : brushedBlocks.entrySet()) {
            block = ((Location) entry.getKey()).getBlock();

            if(block.getType() == Material.AIR) {
                continue;
            }

            block.setType((Material) entry.getValue());
        }
    }

    public static WasteLandFeature getInstance() {
        return instance;
    }

    public void saveGuiItem(int index, ItemStack item) {
        guiItemMap.put(index, item);
    }

    public Map<Integer, ItemStack> getGuiMap() {
        return guiItemMap;
    }

    public Map<Location, Material> getBrushedBlocks() {
        return brushedBlocks;
    }

    public void saveData() {
        // TODO (데이터 I/O 관련 API 학습 필요)
        // guiMap -> 영구저장소

    }

    public void loadData() {
        // TODO (데이터 I/O 관련 API 학습 필요)
        // 영구저장소 -> guiMap
    }

    public void setBrushLevel(ItemStack item, int level) {
        // TODO (데이터 I/O 관련 API 학습 필요)
        // item 에 level 을 부여하여 영구저장소에 저장
    }

    public boolean isConstructionSite(Location location) {
        // TODO (Area 관련 API 필요)
        return true;
    }

    public ItemStack getRandomItem(int level) {
        // <ItemStack, 가중치> 형태의 가중치 Map
        Map<ItemStack, Double> map = new HashMap<>();

        // guiItemMap 에서 ItemStack 을 뽑아 map 으로 put 하며 가중치 부여
        for(Map.Entry<Integer, ItemStack> entry : guiItemMap.entrySet()) {
            // 솔 등급에 따라 확률적으로 받을 수 있는 아이템 거르기
            if(entry.getKey() % 9 >= level) {
                continue;
            }

            map.put(entry.getValue(), getWeight(entry.getKey()));
        }

        Random random = new Random();
        double totalWeight = 0.0D;

        // 총 가중치 합 계산
        for(Double weight : map.values()) {
            totalWeight += weight;
        }

        double randomValue = random.nextDouble() * totalWeight;
        double cumulativeWeight = 0.0D;

        // 가중치에 기반하여 랜덤한 ItemStack 뽑기
        for(Map.Entry<ItemStack, Double> entry : map.entrySet()) {
            cumulativeWeight += entry.getValue();
            if(randomValue <= cumulativeWeight) {
                return entry.getKey();
            }
        }

        // 랜덤으로 ItemStack 을 획득하지 못했을 경우 (이론상 불가능)
        return new ItemStack(Material.AIR);
    }

    public Inventory getGui() {
        return gui.getInventory();
    }

    private double getWeight(int i) {
        return switch (i) {
            case 9, 36 -> 0.30D;
            case 10, 37 -> 0.25D;
            case 11, 38 -> 0.15D;
            case 12, 39 -> 0.10D;
            case 13, 40 -> 0.08D;
            case 14, 41 -> 0.06D;
            case 15, 42 -> 0.04D;
            case 16, 43 -> 0.015D;
            case 17, 44 -> 0.005D;
            default -> 0D;
        };

    }
}

File: src/main/kotlin/io/github/hxxniverse/hobeaktown/feature/wasteland/WastelandFeature.kt
package io.github.hxxniverse.hobeaktown.feature.wasteland

import io.github.hxxniverse.hobeaktown.feature.wasteland.entity.*
import io.github.hxxniverse.hobeaktown.util.base.BaseFeature
import io.github.hxxniverse.hobeaktown.util.database.loggedTransaction
import io.github.hxxniverse.hobeaktown.util.extension.weightedRandomFromList
import org.bukkit.Bukkit
import org.bukkit.Location
import org.bukkit.Material
import org.bukkit.inventory.ItemStack
import org.bukkit.plugin.java.JavaPlugin
import org.jetbrains.exposed.sql.SchemaUtils

class WastelandFeature : BaseFeature {
    override fun onEnable(plugin: JavaPlugin) {
        loggedTransaction {
            SchemaUtils.create(Wastelands, WastelandRewards, WastelandSetups, WastelandSetupRewards, Brushes)
        }
        WastelandCommand().register(plugin)

        Bukkit.getPluginManager().registerEvents(WastelandListener(), plugin)

        loggedTransaction {
            Wasteland.all().forEach { wasteland ->
                if(wasteland.material == Material.GRAVEL) wasteland.material = Material.SUSPICIOUS_GRAVEL
                if(wasteland.material == Material.SAND) wasteland.material = Material.SUSPICIOUS_SAND

                if(wasteland.location.block.type != wasteland.material) {
                    wasteland.location.block.type = wasteland.material
                }
            }
        }
    }

    override fun onDisable(plugin: JavaPlugin) {
    }

    companion object {
        private val WAITING_BLOCK : MutableSet<Location> = mutableSetOf()

        fun addWaiting(loc: Location) {
            WAITING_BLOCK.add(loc)
        }

        fun removeWaiting(loc: Location) {
            WAITING_BLOCK.remove(loc)
        }

        fun randomItem(rewards: Map<Int, ItemStack>, level: Int): ItemStack {
            val map: MutableMap<ItemStack, Int> = mutableMapOf()

            // guiItemMap 에서 ItemStack 을 뽑아 map 으로 put 하며 가중치 부여
            for ((index, itemStack) in rewards) {
                // 현재 레벨에 따라 제외할 인덱스 제외
                if (shouldExclude(index, level)) {
                    continue
                }

                val weight = getWeight(index)
                if (weight > 0) {
                    map[itemStack] = weight
                }
            }

            // map 에 등록된 아이템이 없을 경우 리턴 (이론상 불가능 / 버그 방지)
            if (map.isEmpty()) return ItemStack(Material.AIR)

            // 확장 함수를 사용하여 랜덤한 ItemStack 뽑기
            return map.weightedRandomFromList()
        }

        private fun shouldExclude(index: Int, level: Int): Boolean {
            val base = listOf(9, 36)
            val excludes = mutableSetOf<Int>()

            for (i in 0 until level - 1) {
                excludes.add(base[0] + i)
                excludes.add(base[1] + i)
            }

            return excludes.contains(index)
        }

        private fun getWeight(index: Int): Int {
            return when (index) {
                9, 36 -> 60
                10, 37 -> 50
                11, 38 -> 30
                12, 39 -> 20
                13, 40 -> 16
                14, 41 -> 12
                15, 42 -> 8
                16, 43 -> 3
                17, 44 -> 1
                else -> 0
            }
        }
    }
}